src/hash.cr

```diff
@@ -1010,6 +1022,126 @@ class Hash(K, V)
     updated_entry ? updated_entry.value : yield key
   end
 
+  # Updates the current value of *key* with the value returned by the given block
+  # (the current value is used as input for the block)
+  #
+  # If no key is present, but there's a default value (or default block)
+  # then that default value is used as input for the given block.
+  #
+  # Otherwise it raises.
+  #
+  # It returns the value used as input for the given block
+  # (ie. the old value if key present, or the default value)
+  # ```
+  # h = {"a" => 0, "b" => 1}
+  # h.update("b") { |v| v + 41 }
+  # h["b"] # => 42
+  #
+  # h = Hash(String, Int32).new(40)
+  # h.update("foo") { |v| v + 2 }
+  # h["foo"] # => 42
+  #
+  # h = {} of String => Int32
+  # h.update("a") { 42 } # raises KeyError
+  # ```
+  def update(key : K, &)
```

Comment:
💭 This is a huge method that will be inlined on every call. I'm pretty sure there must be a way to implement this without repeating the entire logic in `upsert`.

For example, the implementation could do this:
- call `find_entry(key)`. It if finds the entry, it updates the value with the given block
- otherwise we need to either insert the new value if there's a block, or raise

What do you think?

Reply:
I agree. As mentioned in the OP, these method share a common structure and it would probably be a good idea to refactor them in a way to avoid huge inlining as well as duplicate code.

Reply:
> 💭 This is a huge method that will be inlined on every call. I'm pretty sure there must be a way to implement this without repeating the entire logic in `upsert`.
> 
> For example, the implementation could do this:
> 
> * call `find_entry(key)`. It if finds the entry, it updates the value with the given block
> * otherwise we need to either insert the new value if there's a block, or raise
> 
> What do you think?

yeah! You're right! Let me share my first implementation (and ideas):
```crystal
def update(key : K, &)
   new_value = if entry = find_entry(key)
                  yield entry.value
               elsif (block = @block) && key.is_a?(K)
                  yield block.call(self, key.as(K))
               else
                  raise KeyError.new "Missing hash key: #{key.inspect}"
               end
   self[key] = new_value
end
``` 
After reading the feature request I thought that we are trying to avoid hash lookups (in this implementation we have the `find_entry` and the `[]=` lookups (`[]=` is just an `upsert`)) 
And so I thought: maybe instead of `[]=` we could use `set_entry(index, new_value)` but then if the `key` is new we don't have an `index`. But that's ok, because we can create an `entry` ... but then we need to take care of all the things involved in creating a new `entry`!! (for example the size of the hash)

So in the end it hit me that I was trying to implement an `upsert` but using a block instead of a value and that's why I based the implementation in `upsert` and `find_entry` (just to avoid using both, as in the first implementation)

Maybe I'm missing some "low-level" methods that we can use to insert a new key (or update)🤔

Reply:
I think what we are missing is a `find_entry_with_index` that's similar to `find_entry` but in addition to returning the `Entry` it also returns the index. With the index, updating the entry value doesn't involve an extra lookup. I think if there's no value and there's a block and we need to insert a new element, doing a double lookup might be okay. After all this will be a one-time thing that will be a bit more expensive than the rest, but then if this is inside a loop or a hot path it will be fast again. And we could eventually optimize it.

Reply:
We don't need an extra method. `#find_entry` could just always return the index alongside the index. It's a private method and we can easily change its signature.

Reply:
ok, I think I understood: not doing an extra lookup on update but not to worry on insert.
I'll do what @straight-shoota mentioned about modifying `#find_entry` and let's see the result 🤓👍